### PR TITLE
Add insert select where support

### PR DESCRIPTION
**Pull request by lunny**
As title.

### CODE REVIEW

File: session_insert_test.go [hunk 1]
```diff
@@ -834,3 +834,62 @@ func TestInsertMap(t *testing.T) {
...
+	}
+
+	assert.NoError(t, prepareEngine())
+	assertSync(t, new(InsertWhere))
+
+	var i = InsertWhere{
+		RepoId: 1,
+		Width:  10,
+		Height: 20,
+		Name:   "trest",
```
Comment (typeless) [inline session_insert_test.go]: typo?

File: session_insert_test.go [hunk 2]
```diff
@@ -834,3 +834,62 @@ func TestInsertMap(t *testing.T) {
...
+	i.Index = 1
+	assert.EqualValues(t, i, j)
+
+	inserted, err = testEngine.Table(new(InsertWhere)).Where("repo_id=?", 1).
+		SetExpr("`index`", "coalesce(MAX(`index`),0)+1").
+		Insert(map[string]interface{}{
+			"repo_id": 1,
+			"width":   20,
+			"height":  40,
+			"name":    "trest2",
```
Comment (typeless) [inline session_insert_test.go]: ditto

File: session_insert_test.go [hunk 3]
```diff
@@ -834,3 +834,62 @@ func TestInsertMap(t *testing.T) {
...
+	assert.EqualValues(t, 1, inserted)
+
+	var j2 InsertWhere
+	has, err = testEngine.ID(2).Get(&j2)
+	assert.NoError(t, err)
+	assert.True(t, has)
+	assert.EqualValues(t, 1, j2.RepoId)
+	assert.EqualValues(t, 20, j2.Width)
+	assert.EqualValues(t, 40, j2.Height)
+	assert.EqualValues(t, "trest2", j2.Name)
```
Comment (typeless) [inline session_insert_test.go]: ditto

Comment (lunny) [reply session_insert_test.go]: It's just a string.

Comment (typeless) [reply session_insert_test.go]: I see. 😄

**Comment by guillep2k** [comment]
I'm a little late for this, but I've tested this with the `Issue.Index` feature in Gitea and it worked perfectly. I'll make a PR to add this over there.

By the way, this is the code generated for SQLITE:

```
2019/08/22 12:21:16 .../xorm/session_raw.go:178:exec() [I] [SQL] INSERT INTO `issue` (`repo_id`, `poster_id`, `original_author`, `original_author_id`, `name`, `content`, `milestone_id`, `priority`, `is_closed`, `is_pull`, `num_comments`, `ref`, `deadline_unix`, `created_unix`, `updated_unix`, `closed_unix`, `is_locked`, `index`) SELECT ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, coalesce(MAX(`index`),0)+1 FROM `issue` WHERE (repo_id=?) []interface {}{1, 1, "", 0, "This is a pull title", "", 0, 0, false, true, 0, "", 0, 1566487276, 1566487276, 0, false, 1} - took: 1.240578ms
```

**Pull request merged**